src/simd/generic.rs

```diff
@@ -68,35 +86,83 @@ pub trait SIMD<
+where
+    ScalarDType: Copy + PartialOrd + AsPrimitive<usize>,
+    SIMDVecDtype: Copy,
+    SIMDMaskDtype: Copy,
+{
     #[inline(always)]
-    unsafe fn _argminmax(data: &[ScalarDType]) -> (usize, usize)
-    where
-        SCALAR: ScalarArgMinMax<ScalarDType>,
-    {
-        argminmax_generic(data, LANE_SIZE, Self::_overflow_safe_core_argminmax)
+    unsafe fn _core_argminmax(arr: &[ScalarDType]) -> (usize, ScalarDType, usize, ScalarDType) {
+        assert_eq!(arr.len() % LANE_SIZE, 0);
+        // Efficient calculation of argmin and argmax together
+        let mut new_index = Self::INITIAL_INDEX;
+        let mut index_low = Self::INITIAL_INDEX;
+        let mut index_high = Self::INITIAL_INDEX;
+
+        let mut arr_ptr = arr.as_ptr(); // Array pointer we will increment in the loop
+        let mut values_low = Self::_mm_loadu(arr_ptr);
+        let mut values_high = Self::_mm_loadu(arr_ptr);
+
+        // This is (40%-5%) slower than the loop below (depending on the data type)
+        // arr.chunks_exact(LANE_SIZE)
+        //     .into_iter()
+        //     .skip(1)
+        //     .for_each(|step| {
+        //         new_index = Self::_mm_add(new_index, increment);
+
+        //         let new_values = Self::_mm_loadu(step.as_ptr());
+
+        //         let lt_mask = Self::_mm_cmplt(new_values, values_low);
+        //         let gt_mask = Self::_mm_cmpgt(new_values, values_high);
+
+        //         index_low = Self::_mm_blendv(index_low, new_index, lt_mask);
+        //         index_high = Self::_mm_blendv(index_high, new_index, gt_mask);
+
+        //         values_low = Self::_mm_blendv(values_low, new_values, lt_mask);
+        //         values_high = Self::_mm_blendv(values_high, new_values, gt_mask);
+        //     });
+
+        for _ in 0..arr.len() / LANE_SIZE - 1 {
+            // Increment the index
+            new_index = Self::_mm_add(new_index, Self::INDEX_INCREMENT);
+            // Load the next chunk of data
+            arr_ptr = arr_ptr.add(LANE_SIZE);
+            // Self::_mm_prefetch(arr_ptr); // Hint to the CPU to prefetch the next chunk of data
+            let new_values = Self::_mm_loadu(arr_ptr);
+
+            // Update the lowest values and index
+            let mask = Self::_mm_cmplt(new_values, values_low);
+            values_low = Self::_mm_blendv(values_low, new_values, mask);
+            index_low = Self::_mm_blendv(index_low, new_index, mask);
+
+            // Update the highest values and index
+            let mask = Self::_mm_cmpgt(new_values, values_high);
+            values_high = Self::_mm_blendv(values_high, new_values, mask);
+            index_high = Self::_mm_blendv(index_high, new_index, mask);
+
+            // 25 is a non-scientific number, but seems to work overall
+            //  => TODO: probably this should be in function of the data type
+            // Self::_mm_prefetch(arr_ptr.add(LANE_SIZE * 25)); // Hint to the CPU to prefetch upcoming data
```

Comment:
is this necessary? surely the predictors can figure out that we're doing this sequentially.

Reply:
Indeed - just wanted to experiment with this, but observed minor / no impact